### PR TITLE
Enhance  GSI to work with two FV3LAM grid orientations.

### PR DESCRIPTION
**Pull request by hu5970**
For this PR:

1) The current GSI can not handle the FV3LAM grid orientation W-E/S-N (SW corner is the first grid point).
The new code was added based on the current GSI FV3LAM interface to handle both grid orientation:

_E-W N-S grid
W-E S-N grid_

2) Bug fix (terrain interpolation) in regional GSI hybrid with dual-resolution ensemble forecast

### CODE REVIEW

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 1]
```diff
@@ -1320,8 +1359,10 @@ subroutine gsi_fv3ncdf_readuv(dynvarsfile,ges_u,ges_v)
              iret=nf90_get_var(gfile_loc,k,uu)
              if(trim(name)=='u'.or.trim(name)=='U') then
                 allocate(temp1(dim(dim_id(1)),dim(dim_id(2)),dim(dim_id(3))))
+                if(.not.grid_reverse_flag) call reverse_grid_r_uv(uu,dim(dim_id(1)),dim(dim_id(2)),dim(dim_id(3)))
```
Comment (WanshuWu-NOAA) [inline src/gsi/gsi_rfv3io_mod.f90:1362]: This is confusing. if not grid reverse then call reverse grid??? 
I suggest that we add comments or reverse the logic.

**Review by WanshuWu-NOAA — APPROVED** [review]

File: src/gsi/gsi_rfv3io_mod.f90 (hunk 1 above)
Comment (WanshuWu-NOAA) [reply src/gsi/gsi_rfv3io_mod.f90:1362]: Thanks for the comments.

**Comment by TingLei-NOAA** [comment]
Hi, Ming,,
I still have some confusion here.
As described by you:
^^

 if(.not.grid_reverse_flag) call reverse_grid_r_uv
!
! and the last input parameter for fv3_h_to_ll is alway true:
!
! call fv3_h_to_ll(u,a,nx,ny,nxa,nya,.true.)

V
However, in the fv3_h_to_ll , with reverse_grid_r_uv= .true.,
So, the wind fields are treated as if they are reversed, that is true,
since, they had been reversed here.
However, those interpolation coefficients like a3dy1(i,j) are still not
"reversed".
Are there any problems here ? or you have already taken care of this?
To me, maybe the simpler way was to call the reverse step again before call
fv3_h_to_ll with the grid_reverse_flag=.false.
Namely
^^
 ! if(.not.grid_reverse_flag) call reverse_grid_r_uv

! and the last input parameter for fv3_h_to_ll is alway true:
! call fv3uv2earth
!added by tl:     if(.not.grid_reverse_flag) call reverse_grid_r_uv   ( to
reverse it to original not-reversed mode)
! call fv3_h_to_ll(u,a,nx,ny,nxa,nya, grid_revserse_flag.)
VV
Let me know if I correctly understand your logics here.
Thanks.
Ting













______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by TingLei-NOAA** [comment]
Ming,
Thanks a lot for your explanation.  I am going to check out your branch for
a closer look to understand your strategy/logics for this development
Mike,
Would you please let me know what is PR ID ?  or how can I check out this
PR to my local machine?
So many thanks.
Ting

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by hu5970** [comment]
@TingLei-NOAA The PR repository is:

https://github.com/hu5970/GSI

master branch

Thanks,
Ming

**Comment by TingLei-NOAA** [comment]
Ming,
Thanks a lot.
ITing

______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Comment by TingLei-NOAA** [comment]
After some reading of Ming's branch, I think I have a better understanding
of this developments. Thanks for Ming's help.
So, in this new option, on the "lowest level" of  the conversion between
fv3 native grid and analysis grids,
the conversion is still done on the basis of original layout : E-W/N-S.
That is why the generate_anl_grid sub is not changed and only the input
lon/lat  are changed to make sure they are in the E-W/N-S layout.
And, the following modifications are to make sure the fields are in
accordance with the setup in subroutine like fv3_h_to_ll.
This approach does work if those details-related modifications are
correctly implemented.
I will approve this PR.
 However, I am wondering , when there are so many details to be modified in
the current implementation, to deal with the different layout in the
background data, should it be considered just to do those
reverse/reverse-back outside of GSI?  I guess even some existing netcdf
tools can do it. The current codes make the code significantly more
difficult to understand (much less readability), while, its goal is just to
deal with the different layout.




______________________________

Ting Lei

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

Ting.Lei@noaa.gov <Joe.Smith@noaa.gov>

301-683-3624

**Review by TingLei-NOAA — APPROVED** [review]
Approved. Comments see the above.

**Review by MichaelLueken-NOAA — COMMENTED** [review]
Hi Ming,

The regression tests were run (both standard and debug) and all tests passed successfully.  I have four minor modifications that will need to be made (adding i_kind to integer declarations (i.e., integer(i_kind) :: i,j,k) and adding i_kind to use kinds, only: r_kind).  Once these changes have been made, the work should be ready for the review committee.

Mike

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 2]
```diff
@@ -2381,5 +2430,51 @@ subroutine gsi_fv3ncdf_write_v1(filename,varname,var,mype_io,add_saved)
...
+end subroutine reverse_grid_r
+
+subroutine reverse_grid_r_uv(grid,nx,ny,nz)
+!
+!  reverse the first two dimension of the array grid
+!
+    use kinds, only: r_kind
+
+    implicit none
+    integer,      intent(in     ) :: nx,ny,nz
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90:2463]: done

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 3]
```diff
@@ -2381,5 +2430,51 @@ subroutine gsi_fv3ncdf_write_v1(filename,varname,var,mype_io,add_saved)
...
+    real(r_kind), intent(inout  ) :: grid(nx,ny,nz)
+    real(r_kind) :: tmp_grid(nx,ny)
+    integer :: i,j,k
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90:2443]: done

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 4]
```diff
@@ -2381,5 +2430,51 @@ subroutine gsi_fv3ncdf_write_v1(filename,varname,var,mype_io,add_saved)
...
 end subroutine gsi_fv3ncdf_write_v1
 
+subroutine reverse_grid_r(grid,nx,ny,nz)
+!
+!  reverse the first two dimension of the array grid
+!
+    use kinds, only: r_kind
+
+    implicit none
+    integer,      intent(in     ) :: nx,ny,nz
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90:2440]: done

File: src/gsi/gsi_rfv3io_mod.f90 [hunk 5]
```diff
@@ -2381,5 +2430,51 @@ subroutine gsi_fv3ncdf_write_v1(filename,varname,var,mype_io,add_saved)
...
+    real(r_kind), intent(inout  ) :: grid(nx,ny,nz)
+    real(r_kind) :: tmp_grid(nx,ny)
+    integer :: i,j,k
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/gsi_rfv3io_mod.f90:2466]: done

**Comment by hu5970** [comment]
@TingLei-NOAA @MichaelLueken-NOAA Thanks for reviewing the code and good comments.

**Comment by MichaelLueken-NOAA** [comment]
Hi Ming,

There should only be one commit associated with a PR.  In your working copy of your fork, please do the following:

1) git reset --soft HEAD~1
2) git commit --amend
3) git push origin master --force

I will then submit PR #102 to the DA review committee.

Thanks,

Mike

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Doen. Let me know if there are other changes needed for sending code out to review.
Thanks,
Ming

**Comment by MichaelLueken-NOAA** [comment]
The due date for this work has past with no feedback from the review committee.  I will now give final approval to these changes and merge them to the NOAA-EMC/GSI master.

**Pull request merged**